Update Compiling Circuits page to include new decompositions system (#7066)
PennyLaneAI/pennylane: doc/introduction/compiling_circuits.rst

.. _intro_ref_compile_circuits:

Quantum compilation

> ===================

Author: Oh, to me "compiling circuits" was closer to catalyst 😅. I'll stick with the original name (if it ain't broke don't fix it)

> Quantum compilation
> ===================

Author: ```suggestion
Compiling circuits
==================
```

`the compilation documentation
<../code/qml_transforms.html#transforms-for-circuit-compilation>`_.

Gate decompositions
-------------------

When compiling a circuit, it is often beneficial to decompose the circuit into a 
set of basis gates. To do this, we can use the :func:`~.pennylane.transforms.decompose` 
function, which enables decomposition of circuits into a set of gates defined either 
by their name, type, or by a set of rules they must follow.

.. note::

    Using :func:`~.pennylane.decompositions.enable_graph` enables PennyLane's new 
    experimental decomposition algorithm. This 

>     new system uses a graph-based approach, which provides better overall performance 

Reviewer: ```suggestion
    new system uses a graph-based approach, which provides better overall resource efficiency 
```

>     and versatility. By default, this new system is *not* enabled globally. 

Reviewer: Should we also include a warning here that this system is still experimental?

>     and versatility. By default, this new system is *not* enabled globally. 

Author: In the .. note:: directive above I have 

>Using :func:`~.pennylane.decompositions.enable_graph` enables PennyLane's new experimental decomposition algorithm.

But I'm happy to be more explicit about the experimental nature of this feature!

>     and versatility. By default, this new system is *not* enabled globally. 

Reviewer: Oh ooops guess I missed it.

`the compilation documentation
<../code/qml_transforms.html#transforms-for-circuit-compilation>`_.

Gate decompositions
-------------------

When compiling a circuit, it is often beneficial to decompose the circuit into a 
set of gates. To do this, we can use the :func:`~.pennylane.transforms.decompose` 
function, which enables decomposition of circuits into a set of gates defined either 
by their name, type, or by a set of rules they must follow.

.. note::

    Using :func:`~.pennylane.decompositions.enable_graph` enables PennyLane's new 
    **experimental** decomposition algorithm (by default, this new system is *not* 
    enabled). This new system uses a graph-based approach, which provides better 
    overall versatility and resource efficiency.

Using a gate set
****************

The example below demonstrates how a three-wire circuit can be decomposed using 
a pre-defined set of gates: 

.. code-block:: python
    
    from pennylane.transforms import decompose
    from functools import partial

    # Works with the new decomposition system, as well.

>     # qml.decompositions.enable_graph()

Reviewer 2: @isaacdevlugt this might be superfluous -- e.g., if this works with graph enabled or disabled, I wouldn't even worry about mentioning this here as it can be confusing!

1: ────╭●─────│─────╭●─────│───T─╰X──T†─╰X─┤     
2: ──H─╰X──T†─╰X──T─╰X──T†─╰X──T──H────────┤ 

.. note::

    Using gate rules with :func:`~.pennylane.decompositions.enable_graph` present 
    is not currently supported. 

Decomposition in stages
***********************

You can use the ``max_expansion`` argument to control the number of decomposition 
stages applied to the circuit. By default, the function will decompose the circuit 
until the desired gate set is reached.

.. note::

    Using the ``max_expansion`` argument with :func:`~.pennylane.decompositions.enable_graph`
    does not do anything, as the new decompositions algorithm is not iterative in 

>     the same way.

Reviewer: I'm sorry if I misled you when you asked me about this. What I meant at the time was that it didn't make sense to use `max_expansion` on the decomposition graph, but you can still control the depth of recursion in the `decompose` transform.

>     the same way.

Author: @astralcai no worries! :) I'll msg you offline about this 👀

>     the same way.

Author: Resolved offline :)

───H†─│────────────╰●───────────────┤  
──────╰●────────────────────────────┤  

Custom Operator Decomposition
-----------------------------

When executing QNodes on a device, PennyLane will automatically decompose gates 
that are unsupported by the device using built-in decomposition rules.

In addition, you can provide *new* decomposition rules to be used, but the behaviour

> and user-interface is different depending on if the new decompositions system is
> enabled.

Reviewer 2: ```suggestion
and user-interface is different depending on if the graph decompositions system is
enabled.
```

I would err towards not calling it 'new decomposition' system as it relies on readers knowing the temporal order of our systems. Rather, better to refer to it based on its feature/algorithm ('graph decompositions'). This also matches the function for turning it on :)

be used.

.. note::

    Device-level custom decompositions **are not applied before other compilation 
    passes (decorators on QNodes)**. For example, the following circuit has ``cancel_inverses`` 
    applied to it, and the device was provided a decomposition for ``qml.CNOT``. 
    The Hadamard gates applied around the ``qml.CNOT`` gate do not get cancelled 
    with those introduced by the custom decomposition.

    .. code-block:: python

        def custom_cnot(wires, **_):
            return [
                qml.H(wires=wires[1]),
                qml.CZ(wires=[wires[0], wires[1]]),
                qml.H(wires=wires[1])
            ]

        dev = qml.device("default.qubit", custom_decomps={qml.CNOT: custom_cnot})

        @qml.transforms.cancel_inverses
        @qml.qnode(dev)
        def circuit():
            qml.H(1)
            qml.CNOT([0, 1])
            qml.H(1)
            return qml.state()

    >>> print(qml.draw(circuit, level="device")())
    0: ───────╭●───────┤  State
    1: ──H──H─╰Z──H──H─┤  State

    To have better control over custom decompositions, consider using the new decompositions

>     system functionality outlined in the next section.

Reviewer 2: same comment here

   :widths: 25 25 25
   :header-rows: 1

   * - 
     - Circuit-level decomposition
     - Device-level decomposition
     - Multiple decompositions per operator
   * - Behaviour with ``qml.decompositions.disable_graph`` (the default setting)
     - No
     - Yes
     - No
   * - Behaviour with ``qml.decompositions.enable_graph``
     - Yes
     - Yes
     - Yes

Behaviour with ``qml.decompositions.disable_graph``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For example, suppose we would like to implement the following QNode:

.. code-block:: python

    def circuit(weights):
        qml.BasicEntanglerLayers(weights, wires=[0, 1, 2])
        return qml.expval(qml.Z(0))

    original_dev = qml.device("default.qubit", wires=3)
    original_qnode = qml.QNode(circuit, original_dev)

>>> weights = np.array([[0.4, 0.5, 0.6]])
>>> print(qml.draw(original_qnode, level="device")(weights))
0: ──RX(0.40)─╭●────╭X─┤  <Z>
1: ──RX(0.50)─╰X─╭●─│──┤     
2: ──RX(0.60)────╰X─╰●─┤     

Now, let's swap out PennyLane's default decomposition of the ``CNOT`` gate into 
``CZ`` and ``Hadamard``. We define the custom decompositions like so, and pass them 
to a device:

.. code-block:: python

>     def custom_cnot(wires, **_):

Reviewer: Right now a quantum function without a resource estimate is not supported.

───H†─│────────────╰●───────────────┤  
──────╰●────────────────────────────┤  

Custom Operator Decomposition
-----------------------------

PennyLane decomposes gates unknown to the device into other, "lower-level" gates. 
As a user, you may want to fine-tune this mechanism. The default behaviour in PennyLane
versus that of the new decompositions system (enabled with ``qml.decomposition.enable_graph``)
differ in the following ways when it comes to injecting custom decompositions for
operators:

> 

Author: I couldn't find anything to properly insert a table in rst lol. Thank you for this!

> 

Reviewer 2: RST also has list tables, which are more convenient https://sublime-and-sphinx-guide.readthedocs.io/en/latest/tables.html

+-------------------+------------------------------+-----------------------------+--------------------------------------+
|                   | QNode-level decomposition    | Device-level decomposition  | Multiple decompositions per operator |
+===================+==============================+=============================+======================================+
| ``disable_graph`` | No                           | Yes                         | No                                   |
+-------------------+------------------------------+-----------------------------+--------------------------------------+
| ``enable_graph``  | Yes                          | Yes                         | Yes                                  |

> +-------------------+------------------------------+-----------------------------+--------------------------------------+

Author: Yeah this was just an experiment. Wanted to see how it would render in the docs and get a better overall picture of it in context. I just couldn't figure out how to create a table lol

    passes (decorators on QNodes)**. For example, the following circuit has ``cancel_inverses`` 
    applied to it, and the device was provided a decomposition for ``qml.CNOT``. 
    The Hadamard gates applied around the ``qml.CNOT`` gate do not get cancelled 
    with those introduced by the custom decomposition.

    .. code-block:: python

        def custom_cnot(wires, **_):
            return [
                qml.H(wires=wires[1]),
                qml.CZ(wires=[wires[0], wires[1]]),
                qml.H(wires=wires[1])
            ]

        dev = qml.device("default.qubit", custom_decomps={qml.CNOT: custom_cnot})

        @qml.transforms.cancel_inverses
        @qml.qnode(dev)
        def circuit():
            qml.H(1)
            qml.CNOT([0, 1])
            qml.H(1)
            return qml.state()

    >>> print(qml.draw(circuit, level="device")())
    0: ───────╭●───────┤  State
    1: ──H──H─╰Z──H──H─┤  State

    To have better control over custom decompositions, consider using the new decompositions
    system functionality outlined in the next section.

Behaviour with qml.decomposition.enable_graph

> *********************************************

Reviewer 2: this section is great!

───H†─│────────────╰●───────────────┤  
──────╰●────────────────────────────┤  

Custom Operator Decomposition
-----------------------------

When executing QNodes on a device, PennyLane will automatically decompose
gates that are unsupported by the device using built-in decomposition rules.

In addition, you can provide custom decomposition rules to be used.

The default behaviour in PennyLane
versus that of the new decompositions system (enabled with ``qml.decomposition.enable_graph``)
differ in the following ways when it comes to injecting custom decompositions for
operators:

+-------------------+------------------------------+-----------------------------+--------------------------------------+

> |                   | QNode-level decomposition    | Device-level decomposition  | Multiple decompositions per operator |

Reviewer: What do you mean by QNode-level and Device-level exactly?

+===================+==============================+=============================+======================================+
| ``disable_graph`` | No                           | Yes                         | No                                   |
+-------------------+------------------------------+-----------------------------+--------------------------------------+
| ``enable_graph``  | Yes                          | Yes                         | Yes                                  |
+-------------------+------------------------------+-----------------------------+--------------------------------------+

> Behaviour with qml.decomposition.disable_graph

Reviewer: Since the new system is disabled by default, this is just the default behaviour. Maybe instead of phrasing it like enable_graph vs. disable_graph, phrase it like current behaviour / experimental behaviour with graph enabled?

    with those introduced by the custom decomposition.

    .. code-block:: python

        def custom_cnot(wires, **_):
            return [
                qml.H(wires=wires[1]),
                qml.CZ(wires=[wires[0], wires[1]]),
                qml.H(wires=wires[1])
            ]

        dev = qml.device("default.qubit", custom_decomps={qml.CNOT: custom_cnot})

        @qml.transforms.cancel_inverses
        @qml.qnode(dev)
        def circuit():
            qml.H(1)
            qml.CNOT([0, 1])
            qml.H(1)
            return qml.state()

    >>> print(qml.draw(circuit, level="device")())
    0: ───────╭●───────┤  State
    1: ──H──H─╰Z──H──H─┤  State

    To have better control over custom decompositions, consider using the new decompositions
    system functionality outlined in the next section.

Behaviour with qml.decomposition.enable_graph
*********************************************

With the new graph-based system enabled, custom decompositions for operators in PennyLane
can be added in a few ways depending on the application. 

The :func:`~.pennylane.transforms.decompose` transform offers the ability to inject
custom decompositions via two keyword arguments:

> * ``fixed_decomps``: any operator with a decomposition listed here will automatically 
>   be chosen by the new algorithm, regardless of how efficient it may or may not 
>   be.
> * ``alt_decomps``: any operator with a decomposition list here is added as a possible 
>   decomposition the algorithm can choose based on its resources.

Author: ```suggestion
* ``fixed_decomps``: any decomposition for an operator type here will automatically 
  be chosen by the new algorithm, regardless of how resource efficient it may or 
  may not be.
* ``alt_decomps``: any decompositions for an operator type list here are added as 
  a *possible* decomposition rules the algorithm can choose based on its resource 
  efficiency.
```

> Both keyword arguments above require a dictionary mapping PennyLane operators to 

Reviewer: ```suggestion
Both keyword arguments above require a dictionary mapping PennyLane operator types to 
```

custom decompositions. Creating custom decompositions that the system can use involves 

> a PennyLane quantum function that represents the decomposition and a registering 
> resource data to it that tracks gate counts in the custom decomposition via :func:`~.pennylane.register_resources`.

Reviewer: ```suggestion
a PennyLane quantum function that represents the decomposition and a declaration of how much resource it requires in terms of gate counts, via :func:`~.pennylane.register_resources`.
```

Author: Nice! Slight amendment based on your suggestion

```suggestion
a PennyLane quantum function that represents the decomposition, and a declaration of its resource requirements (gate counts) via :func:`~.pennylane.register_resources`.
```

Consider this example where we add a fixed decomposition to ``CNOT`` gates:

.. code-block:: python

    @qml.register_resources({qml.H: 2, qml.CZ: 1})
    def my_cnot(wires, **__):
        qml.H(wires=wires[1])
        qml.CZ(wires=wires)
        qml.H(wires=wires[1])

The :func:`~.pennylane.register_resources` accepts a dictionary mapping operators 
within the custom decomposition to the number of times they occur in the decomposition. 
With the resources registered, this can be used with ``fixed_decomps`` or ``alt_decomps``:

.. code-block:: python

    @partial(qml.transforms.decompose, fixed_decomps={qml.CNOT: my_cnot})
    @qml.qnode(qml.device("default.qubit"))
    def circuit():
        qml.CNOT(wires=[0, 1])
        return qml.state()

>>> print(qml.draw(circuit, level="device")())
0: ────╭●────┤  State
1: ──H─╰Z──H─┤  State

> Note that the ``alt_decomps`` argument can handle multiple options per operator:

Reviewer: ```suggestion
Note that the ``alt_decomps`` argument can handle multiple alternatives per operator:
```
